Add sdk for Content Safety

mengaims opened the pull request:
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

mengaims committed:
Change namespace for project carnegie

mengaims committed:
rename from ProjectCarnegie to ContentSafety

mengaims committed:
Fix build errors

mengaims committed:
Update tests and samples for ContentSafety

mengaims committed:
Merge branch 'Azure:main' into carnegie

pallavit commented:
/cc: @heaths for review/service buddy.

heaths commented on sdk/contentsafety/Azure.AI.ContentSafety/tsp-location.yaml:
```diff
@@ -0,0 +1,4 @@
+# commit: 345dd133648fea2046f2fb484111ab4dba5ff34b
+# repo: https://github.com/mengaims/azure-rest-api-specs-pr/
+directory: specification/cognitiveservices/ContentSafety
+spec-root-dir: D:\src\mengaims\azure-rest-api-specs-pr
```
We shouldn't be putting local paths that won't resolve on most machines in the repo. @m-nash is this expected?

heaths replied:
The fact is that it won't work on most people's machines seems to indicate it's not even necessary.

m-nash replied:
Yeah it won't pass CI either like this.

mengaims replied:
Removed this file.

heaths commented on sdk/contentsafety/Azure.AI.ContentSafety/README.md:
```diff
@@ -0,0 +1,91 @@
+# Azure ContentSafety client library for .NET
+
+This section should give out brief introduction of the client library.
```
Should this be filled out before being merged for the first time, or is another PR coming?

mengaims replied:
This will be included in this PR, WIP.

heaths commented on sdk/contentsafety/Azure.AI.ContentSafety/api/Azure.AI.netstandard2.0.cs:
```diff
@@ -0,0 +1,17 @@
+namespace Azure.AI
+{
+    public partial class Client
```
This name is too vague. You should set the full client name e.g., `ContentSafetyClient`, in your client tsp.

heaths replied:
All other type names with the client name - options, tests, etc. - should be similarly named.

heaths replied:
Oh, seems like the actual generated code may be correct. You need to run `eng/scripts/CodeChecks.ps1 contentsafety` to update all this.

mengaims replied:
@heaths I think this problem is related to the bug I encountered when generating sdk project following this [guidance](https://github.com/Azure/azure-sdk-for-net/blob/main/doc/DataPlaneCodeGeneration/AzureSDKPackage_Setup.md)
What this step produced all have the name Azure.AI instead of Azure.AI.ContentSafety. I manually fixed most of the names and missed this one.
I'm wondering what's your suggestion here? Continue fixing the name manually on my original project or ask someone's help to fix this issue in PS script?
You can reproduce it by running: 
.\eng\scripts\automation\Invoke-TypeSpecDataPlaneGenerateSDKPackage.ps1  -sdkFolder D:\src\mengaims\azure-sdk-for-net\sdk\contentsafety\Azure.AI.ContentSafety -typespecSpecDirectory specification\cognitiveservices\ContentSafety -commit 462e8d9327ea0dfc77e7df76933b34b3d489953e -repo Azure\azure-rest-api-specs-pr
And I pushed the output sdk project [here](https://github.com/mengaims/azure-sdk-for-net/tree/acs/sdk/contentsafety).

mengaims replied:
Fixed manually.

heaths commented on sdk/contentsafety/Azure.AI.ContentSafety/api/Azure.AI.netstandard2.0.cs:
```diff
@@ -0,0 +1,17 @@
+namespace Azure.AI
```
You need to have a service area under the topic. See https://azure.github.io/azure-sdk/general_design.html#general-namespaces-naming for guidelines.

mengaims replied:
@heaths I think this problem is related to the bug I encountered when generating sdk project following this [guidance](https://github.com/Azure/azure-sdk-for-net/blob/main/doc/DataPlaneCodeGeneration/AzureSDKPackage_Setup.md)
What this step produced all have the name Azure.AI instead of Azure.AI.ContentSafety. I manually fixed most of the names and missed this one.
I'm wondering what's your suggestion here? Since running eng/scripts/CodeChecks.ps1 contentsafety would not fix this problem. Do I need to fix manually? Or do you know how to fix this issue in script?
You can reproduce it by running: 
.\eng\scripts\automation\Invoke-TypeSpecDataPlaneGenerateSDKPackage.ps1  -sdkFolder D:\src\mengaims\azure-sdk-for-net\sdk\contentsafety\Azure.AI.ContentSafety -typespecSpecDirectory specification\cognitiveservices\ContentSafety -commit 462e8d9327ea0dfc77e7df76933b34b3d489953e -repo Azure\azure-rest-api-specs-pr
And I pushed the output sdk project [here](https://github.com/mengaims/azure-sdk-for-net/tree/acs/sdk/contentsafety).

mengaims replied:
I've done some manual fix and ran eng/scripts/CodeChecks.ps1 contentsafety to update. Could you check if the new file Azure.AI.ContentSafety.netstandard2.0.cs is valid for you?

mengaims replied:
Fixed namespace manually

heaths commented on sdk/contentsafety/Azure.AI.ContentSafety/api/Azure.AI.netstandard2.0.cs:
```diff
@@ -0,0 +1,17 @@
+        protected Client() { }
+        public Client(string endpoint, Azure.Core.TokenCredential credential, Azure.AI.ClientOptions options = null) { }
+        public virtual Azure.Core.Pipeline.HttpPipeline Pipeline { get { throw null; } }
+    }
```
What are the operations? This client seems to do nothing.

mengaims replied:
Merge the discussion about operation missing to one thread and close this one.

heaths commented on sdk/contentsafety/Azure.AI.ContentSafety/samples/README.md:
This and subordinate samples also need to be filled in, at least eventually. Another PR forthcoming?

mengaims replied:
@heaths  I placed the samples in tests/samples, and removed this one. Does it work for you?

mengaims committed:
Merge branch 'Azure:main' into carnegie

mengaims committed:
update generated code

mengaims committed:
Rename Azure.AI.netstandard2.0.cs

mengaims committed:
Update sample2

heaths commented on sdk/contentsafety/Azure.AI.ContentSafety/api/Azure.AI.ContentSafety.netstandard2.0.cs:
```diff
@@ -0,0 +1,17 @@
+        protected ContentSafetyClient() { }
+        public ContentSafetyClient(string endpoint, Azure.Core.TokenCredential credential, Azure.AI.ClientOptions options = null) { }
+        public virtual Azure.Core.Pipeline.HttpPipeline Pipeline { get { throw null; } }
+    }
```
This still shows no methods. Did you run _eng/scripts/CodeChecks.ps1 contentsafety_ and commit?

mengaims replied:
Yep I've already ran eng/scripts/CodeChecks.ps

mengaims replied:
Hi @heaths, do you have any updated suggestion for this problem? I've created an issue reporting this bug: https://github.com/Azure/autorest.csharp/issues/3342. We plan to release this sdk in //build, and target merge this PR by 5/12, would appreciate if you could help pushing the fix of this issue.

heaths replied:
This should have all public members in https://github.com/Azure/azure-sdk-for-net/pull/34933/files#diff-5ba963678ef26de570a5282ed29b4864b1ecdb8a94084d4f74a50f5050c97305 and all your public models defined in this file. Does the script show any errors when you run it? Does this project even build?

As far as releasing, you're still missing a README, samples, and more - much of which is required to release. We have validations in place that make sure your README has appropriate content, as well as CHANGELOGs. The APIs also have to be reviewed, and if this file isn't getting populated there's nothing for the architects to review and approve.

mengaims replied:
Hi @heaths, thanks for the response! I have uploaded tests and samples, and will work on the README and CHANGELOG before PR get merged.
This SDK has already been reviewed by SDK architects, please refer to this [link](https://apiview.dev/Assemblies/Review/d1fee5210fa34184bca6003f087171e3).
![image](https://user-images.githubusercontent.com/50131889/236097456-be64a89d-9577-43e2-9664-0e27246dcfcb.png)
We've already got approval for the first release. Ted said for the first release only the approval of package name is needed, and we will collect customer feedbacks during the Public Preview.
I'm wondering if we could have a quick sync this week about the remaining items before package release. And maybe I can show you the problem I met during code generation then.

heaths replied:
You need to fix those for builds to work. Looks like your _.sln_ file references a non-existent project - probably just the wrong path. You can edit it by hand. You should be able to run `dotnet build` or even `dotnet test` in your solution directly cleanly.

@m-nash I know some edits were made after the code was generated, but I wonder if the template needs to be fixed. Perhaps the _.sln_ doesn't have the right template like it once did.

heaths replied:
The _stress/*.csproj_ project is missing. You should open a tracking issue to add it back, but for now I can remove it from the _.sln_.

heaths replied:
That's not the problem I'm seeing. When I checked out and build your PR, there are compilation errors e.g.,

> error CS0246: T
he type or namespace name 'IUtf8JsonSerializable' could not be found (are you missing a using directive or an assembly reference?) [C:\src\az
ure-sdk-for-net\sdk\contentsafety\Azure.AI.ContentSafety\src\Azure.AI.ContentSafety.csproj::TargetFramework=netstandard2.0]

@m-nash this is a generated client, but what are they missing that this isn't compiling?

Also, since there are models shouldn't be there client teaspoons? I don't see those in this PR.

mengaims replied:
@heaths I've tried some mitigation of this issue: https://github.com/Azure/autorest.csharp/issues/3341, and after this mitigated, the operations seem generated successfully. I pushed it here in another branch: https://github.com/mengaims/azure-sdk-for-net/blob/carnegie-test/sdk/contentsafety/Azure.AI.ContentSafety/api/Azure.AI.ContentSafety.netstandard2.0.cs
I'm confirming with Crystal whether my mitigation is valid, if so, I think we could replace with latest generated code.

Btw, I once encountered the same build error of IUtf8JsonSerializable, but after I close the solution and reopen again it disappeared, you could give it a try.

heaths replied:
Why did you push it to another branch? This PR should've been updated. We can't merge it until it's in a buildable state.

m-nash replied:
You get models with tsp without a client.tsp they just use all default settings.  Client.tsp is only needed for customization.  The reason this `IUtf8JsonSerializable` error is happenig is because there is no tsp-location.yaml file checked in with this PR that I could see.

mengaims committed:
Fix image sample

mengaims committed:
Add blocklist management sample

mengaims committed:
Merge branch 'Azure:main' into carnegie

mengaims committed:
Suppress AZC0002

mengaims committed:
Fix GetTextBlocklistItems return type

mengaims committed:
Fix CreateOrUpdateTextBlocklist manually

mengaims committed:
Remove useless samples

mengaims committed:
Resolve comments

mengaims commented:
> @mengaims could you please update the codeowners file. We always recommend putting 2 owners in the list - https://github.com/Azure/azure-sdk-for-net/blob/main/.github/CODEOWNERS

@pallavit Updated, please check the latest code in PR. Btw, do I need any other changes like changing the name of PR?

mengaims committed:
Update sample and test recordings

mengaims committed:
Update CI and Changelog

mengaims commented:
@heaths I've already updated following content in this PR:

1. Tests and test recordings
2. Samples
3. Changelog
4. Ci.yml

Could you help have one round of review of them and trigger the CI pipeline run?

And here are the items WIP:

1. Readme
2. Operations generation issue in Azure.AI.ContentSafety.netstandard2.0.cs (We could have an online discussion if you could attend the sync I sent)

Please let me know if I have other things missing.

heaths commented:
> Could you help have one round of review of them and trigger the CI pipeline run?

@m-nash should the onboarding tool be creating the pipelines automatically? @mengaims there's no pipeline for me to approve, but one was not created which seems the onboarding wizard should've taken care of. Or was there a step skipped in documentation? Looking through our wiki docs, it seems this should've been created automatically. I'll ping a few people offline for clarification.

heaths commented:
/azp run prepare-pipelines

m-nash commented:
/azp run net - contentsafety - ci

m-nash commented:
Now that the CI is hooked in it will start to report all the errors.

One thing I see is we are missing the tsp-location.yaml file which was used to do the dotnet build /t:GenerateCode.  The CI will complain about this.

mengaims commented:
> Now that the CI is hooked in it will start to report all the errors.
> 
> One thing I see is we are missing the tsp-location.yaml file which was used to do the dotnet build /t:GenerateCode. The CI will complain about this.

@m-nash , thanks for the reminding! I have a question here, the CI will only check the existence of this file or will try run dotnet build /t:GenerateCode according to it? Since I've met some errors when using remote branch in this file, after mitigate with referencing to local file, it works.

heaths commented:
@mengaims you need to run `dotnet build /t:GenerateCode` locally. See https://aka.ms/azsdk/dpcodegen/net. You also need to check in all the files including the client tsp.

mengaims commented:
> @mengaims you need to run `dotnet build /t:GenerateCode` locally. See https://aka.ms/azsdk/dpcodegen/net. You also need to check in all the files including the client tsp.

Yes I've ran it locally.

m-nash committed:
run export api
comment out unused snippets
update test to pass base64 string to binarydata
add tsp-location and gen library

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into carnegie

m-nash committed:
regen with latest client

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.AI.ContentSafety](https://apiview.dev/Assemblies/Review/dd3a9609479549b5a1e6581c556792c6)

m-nash committed:
fix broken links

m-nash committed:
update nuget since the package hasn't shipped yet

m-nash committed:
fix link in samples readme

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into carnegie

m-nash committed:
regen with latest generator

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into carnegie

m-nash committed:
regen with latest and update test to use the image.jpg bytes vs hard coded base64 string
